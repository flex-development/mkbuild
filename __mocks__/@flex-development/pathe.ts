/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
