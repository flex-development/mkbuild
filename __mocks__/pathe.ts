/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
