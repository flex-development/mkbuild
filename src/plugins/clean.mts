/**
 * @file Plugins - clean
 * @module mkbuild/plugins/clean
 */

import pathe from '@flex-development/pathe'
import { ok } from 'devlop'
import type {
  NormalizedOutputOptions,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:clean'

/**
 * Cleaned directory cache.
 *
 * @internal
 *
 * @const {Set<string>} directories
 */
const directories: Set<string> = new Set<string>()

plugin.PLUGIN_NAME = PLUGIN_NAME
plugin.directories = directories

export default plugin

/**
 * Create an output directory cleaner plugin.
 *
 * @internal
 *
 * @see {@linkcode Plugin}
 *
 * @this {void}
 *
 * @return {Plugin}
 *  Output directory cleaner plugin
 */
function plugin(this: void): Plugin {
  return {
    name: PLUGIN_NAME,
    renderStart: { handler: renderStart, order: 'pre', sequential: true }
  }

  /**
   * Remove and recreate an output directory.
   *
   * @see https://rollupjs.org/plugin-development/#renderstart
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} output
   *  Normalized output options
   * @return {Promise<undefined>}
   */
  async function renderStart(
    this: PluginContext,
    output: NormalizedOutputOptions
  ): Promise<undefined> {
    ok(typeof output.clean === 'boolean', 'expected `output.clean`')
    ok(typeof output.fs === 'object', 'expected `output.fs`')
    ok(typeof output.root === 'string', 'expected `output.root`')
    ok(typeof output.write === 'boolean', 'expected `output.write`')

    if (
      output.clean &&
      output.root &&
      output.write &&
      typeof output.dir === 'string'
    ) {
      directories.add(output.root)

      /**
       * Absolute path to output directory.
       *
       * @const {string} path
       */
      const path: string = pathe.join(output.root, output.dir)

      if (!directories.has(path)) {
        await output.fs.rm(path, { force: true, recursive: true })
        await output.fs.mkdir(path, { recursive: true })
      }

      directories.add(path)
    }

    return void this
  }
}
