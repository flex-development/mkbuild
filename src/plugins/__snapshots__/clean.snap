// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:plugins/clean > should return plugin 1`] = `
{
  "name": "mkbuild:clean",
  "renderStart": {
    "handler": [Function renderStart],
    "order": "pre",
    "sequential": true,
  },
}
`;
